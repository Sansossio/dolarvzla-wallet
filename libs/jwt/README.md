# Environment variables

```env
# Jwt configuration
SECURITY_JWT_SECRET=secret
SECURITY_JWT_EXPIRATION=1h

# Encrypt token using AES
JWT_ENCRYPTION_SECRET=Y4p9CAaSJzVkILxJGjXdn08ju2824mr2
```
