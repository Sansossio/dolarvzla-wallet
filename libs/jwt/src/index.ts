/* Module */
export * from './lib/jwt.module'
/* Service */
export * from './lib/service/jwt.crypto.service'
/* type */
export * from './lib/type/register-async-jwt.type'
export * from './lib/type/register-jwt.type'
export * from './lib/type/request.type'
export * from './lib/type/token.type'
