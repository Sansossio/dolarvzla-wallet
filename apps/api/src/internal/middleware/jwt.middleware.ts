import { Injectable, NestMiddleware } from '@nestjs/common'
import { Request } from 'express'
import { UserEntity } from '@dolarvzla-wallet/models'
import { JwtCryptoService } from '@dolarvzla-wallet/jwt'

@Injectable()
export class JwtMiddleware implements NestMiddleware {
  constructor (
    private readonly service: JwtCryptoService
  ) {}

  async use (req: Request & { user: UserEntity }, _: Response, next: Function) {
    const token: string = req.headers.authorization

    if (typeof token === 'string') {
      try {
        const data: { sub: string } = this.service.decode(token)
        req.tokenData = req.tokenData || {}

        req.tokenData.sub = data.sub
      } catch (e) {
        // Ignore error
        console.log()
      }
    }

    next()
  }
}
