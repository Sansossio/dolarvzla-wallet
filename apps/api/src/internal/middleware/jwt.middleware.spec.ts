import { JwtMiddleware } from './jwt.middleware'

describe('Jwt Middleware', () => {
  const next: any = () => true
  const jwtService = {
    decode: jest.fn()
  }

  const service = new JwtMiddleware(
    jwtService as any
  )

  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('Use', () => {
    it('should add objects to req', async () => {
      jwtService.decode.mockImplementation(() => ({ sub: 'ababe672-36d2-4d45-839a-34babe67289e' }))

      const req: any = {
        headers: {
          authorization: 'token'
        }
      }

      await service.use(req, null, next)

      expect(req).toMatchSnapshot()
    })

    it('should add objects to req when casting userId (sub attribute)', async () => {
      jwtService.decode.mockImplementation(() => ({ sub: '42' }))

      const req: any = {
        headers: {
          authorization: 'token'
        }
      }

      await service.use(req, null, next)

      expect(req).toMatchSnapshot()
    })

    it('should ignore implementation when token is not string', async () => {
      jwtService.decode.mockImplementation(() => (true))

      const req: any = {
        headers: {
          authorization: true
        }
      }

      await service.use(req, null, next)
      expect(req).toMatchSnapshot()
    })

    it('should ignore error when userId (sub attribute) is not valid', async () => {
      jwtService.decode.mockImplementation(() => ({ sub: 'wrong' }))

      const req: any = {
        headers: {
          authorization: 'token'
        }
      }
      await service.use(req, null, next)
      expect(req).toMatchSnapshot()
    })

    it('should ignore error on implementation', async () => {
      jwtService.decode.mockImplementation(() => {
        throw new Error()
      })

      const req: any = {
        headers: {
          authorization: 'token'
        }
      }

      await service.use(req, null, next)
      expect(req).toMatchSnapshot()
    })
  })
})
