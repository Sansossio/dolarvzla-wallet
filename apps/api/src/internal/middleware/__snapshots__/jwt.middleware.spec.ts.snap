// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jwt Middleware Use should add objects to req 1`] = `
Object {
  "headers": Object {
    "authorization": "token",
  },
  "token": "token",
  "tokenData": Object {
    "sub": "ababe672-36d2-4d45-839a-34babe67289e",
  },
}
`;

exports[`Jwt Middleware Use should add objects to req when casting userId (sub attribute) 1`] = `
Object {
  "headers": Object {
    "authorization": "token",
  },
  "token": "token",
  "tokenData": Object {
    "sub": "ababe672-36d2-4d45-839a-34babe67289e",
  },
}
`;

exports[`Jwt Middleware Use should ignore error on implementation 1`] = `
Object {
  "headers": Object {
    "authorization": "token",
  },
}
`;

exports[`Jwt Middleware Use should ignore error when userId (sub attribute) is not valid 1`] = `
Object {
  "headers": Object {
    "authorization": "token",
  },
}
`;

exports[`Jwt Middleware Use should ignore implementation when token is not string 1`] = `
Object {
  "headers": Object {
    "authorization": true,
  },
}
`;
